shell/platform/windows/egl/surface.cc

```diff
@@ -0,0 +1,66 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/egl/surface.h"
+
+#include "flutter/shell/platform/windows/egl/egl.h"
+
+namespace flutter {
+namespace egl {
+
+Surface::Surface(EGLDisplay display, EGLContext context, EGLSurface surface)
+    : display_(display), context_(context), surface_(surface) {}
+
+Surface::~Surface() {
+  Destroy();
+}
+
+bool Surface::IsValid() const {
+  return is_valid_;
+}
+
+bool Surface::Destroy() {
+  if (surface_ != EGL_NO_SURFACE) {
+    // Ensure the surface is not current before destroying it.
+    if (::eglMakeCurrent(display_, EGL_NO_SURFACE, EGL_NO_SURFACE,
+                         EGL_NO_CONTEXT) != EGL_TRUE) {
+      WINDOWS_LOG_EGL_ERROR;
+      return false;
+    }
+
+    if (::eglDestroySurface(display_, surface_) != EGL_TRUE) {
+      WINDOWS_LOG_EGL_ERROR;
+      return false;
+    }
+  }
```

Comment:
Are we enforcing anywhere which thread this class is called from? If not, should we add debug build FML checks to avoid the potential for race conditions if someone down the road changes things?

e.g. if this were called from two threads and the first thread were just about to start executing the code at line 26 as the second finishes the executing the whole block ending line 36, things would be sad

Reply:
We're not enforcing which thread this is called from yet, but we really should given OpenGL's behavior is heavily dependent on the current thread.

There's some surface operations that happen on both the platform and raster threads:

1. Make a surface current: on the platform thread for initialization, on the raster thread otherwise
1. Destroying a surface: on raster thread for resizing, or on platform thread for engine shutdown

In other words, this layer doesn't know enough to assert the current thread is expected. We likely will want to sprinkle these kinds of assertions in the view and compositor abstractions. I'll need to think more on this :)